### ISSUE TITLE
Two of CookieStoreTest get ActionDispatch::Session::SessionRestoreError since #43028

### ISSUE THREAD

**Issue by yahonda** [rails/rails#43033]
Here are steps to reproduce https://buildkite.com/rails/rails/builds/80302 since https://github.com/rails/rails/pull/43028

### Steps to reproduce

```ruby
cd actionpack
bin/test test/dispatch/session/cookie_store_test.rb -n test_deserializes_unloaded_classes_on_get_id
bin/test test/dispatch/session/cookie_store_test.rb -n test_deserializes_unloaded_classes_on_get_value
```
### Expected behavior
It should work.

### Actual behavior

Two of CookieStoreTest get ActionDispatch::Session::SessionRestoreError since #43028

```ruby
% bin/test test/dispatch/session/cookie_store_test.rb -n test_deserializes_unloaded_classes_on_get_id
Run options: -n test_deserializes_unloaded_classes_on_get_id --seed 20545

# Running:

E

Error:
CookieStoreTest#test_deserializes_unloaded_classes_on_get_id:
ActionDispatch::Session::SessionRestoreError: Session contains objects whose class definition isn't available.
Remember to require the classes for all objects kept in the session.
(Original exception: uninitialized constant SessionAutoloadTest [NameError])

    /Users/yahonda/src/github.com/rails/rails/activesupport/lib/active_support/inflector/methods.rb:280:in `constantize'
    /Users/yahonda/src/github.com/rails/rails/activesupport/lib/active_support/core_ext/string/inflections.rb:74:in `constantize'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/middleware/session/abstract_store.rb:58:in `rescue in stale_session_check!'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/middleware/session/abstract_store.rb:52:in `stale_session_check!'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/middleware/session/cookie_store.rb:88:in `block in unpacked_cookie_data'
    /Users/yahonda/.rbenv/versions/3.0.2/lib/ruby/gems/3.0.0/gems/rack-2.2.3/lib/rack/request.rb:69:in `fetch'
    /Users/yahonda/.rbenv/versions/3.0.2/lib/ruby/gems/3.0.0/gems/rack-2.2.3/lib/rack/request.rb:69:in `fetch_header'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/middleware/session/cookie_store.rb:87:in `unpacked_cookie_data'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/middleware/session/cookie_store.rb:81:in `block in extract_session_id'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/middleware/session/abstract_store.rb:53:in `stale_session_check!'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/middleware/session/cookie_store.rb:80:in `extract_session_id'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/request/session.rb:65:in `block in id'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/request/session.rb:64:in `fetch'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/request/session.rb:64:in `id'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/request/session.rb:84:in `id'
    /Users/yahonda/src/github.com/rails/rails/actionpack/test/dispatch/session/cookie_store_test.rb:39:in `get_session_id'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_controller/metal/basic_implicit_render.rb:6:in `send_action'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/abstract_controller/base.rb:214:in `process_action'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_controller/metal/rendering.rb:53:in `process_action'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/abstract_controller/callbacks.rb:221:in `block in process_action'
    /Users/yahonda/src/github.com/rails/rails/activesupport/lib/active_support/callbacks.rb:99:in `run_callbacks'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/abstract_controller/callbacks.rb:220:in `process_action'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_controller/metal/rescue.rb:22:in `process_action'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_controller/metal/instrumentation.rb:65:in `block in process_action'
    /Users/yahonda/src/github.com/rails/rails/activesupport/lib/active_support/notifications.rb:207:in `block in instrument'
    /Users/yahonda/src/github.com/rails/rails/activesupport/lib/active_support/notifications/instrumenter.rb:24:in `instrument'
    /Users/yahonda/src/github.com/rails/rails/activesupport/lib/active_support/notifications.rb:207:in `instrument'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_controller/metal/instrumentation.rb:64:in `process_action'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_controller/metal/params_wrapper.rb:250:in `process_action'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/abstract_controller/base.rb:151:in `process'
    /Users/yahonda/src/github.com/rails/rails/actionview/lib/action_view/rendering.rb:39:in `process'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_controller/metal.rb:188:in `dispatch'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_controller/metal.rb:251:in `dispatch'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/routing/route_set.rb:49:in `dispatch'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/routing/route_set.rb:32:in `serve'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/journey/router.rb:50:in `block in serve'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/journey/router.rb:32:in `each'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/journey/router.rb:32:in `serve'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/routing/route_set.rb:845:in `call'
    /Users/yahonda/.rbenv/versions/3.0.2/lib/ruby/gems/3.0.0/gems/rack-2.2.3/lib/rack/session/abstract/id.rb:266:in `context'
    /Users/yahonda/.rbenv/versions/3.0.2/lib/ruby/gems/3.0.0/gems/rack-2.2.3/lib/rack/session/abstract/id.rb:260:in `call'
    /Users/yahonda/.rbenv/versions/3.0.2/lib/ruby/gems/3.0.0/gems/rack-2.2.3/lib/rack/head.rb:12:in `call'
    /Users/yahonda/.rbenv/versions/3.0.2/lib/ruby/gems/3.0.0/gems/rack-2.2.3/lib/rack/method_override.rb:24:in `call'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/middleware/cookies.rb:693:in `call'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/middleware/callbacks.rb:27:in `block in call'
    /Users/yahonda/src/github.com/rails/rails/activesupport/lib/active_support/callbacks.rb:99:in `run_callbacks'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/middleware/callbacks.rb:26:in `call'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/middleware/actionable_exceptions.rb:17:in `call'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/middleware/debug_exceptions.rb:28:in `call'
    /Users/yahonda/src/github.com/rails/rails/actionpack/test/abstract_unit.rb:96:in `call'
    /Users/yahonda/.rbenv/versions/3.0.2/lib/ruby/gems/3.0.0/gems/rack-test-1.1.0/lib/rack/mock_session.rb:29:in `request'
    /Users/yahonda/.rbenv/versions/3.0.2/lib/ruby/gems/3.0.0/gems/rack-test-1.1.0/lib/rack/test.rb:266:in `process_request'
    /Users/yahonda/.rbenv/versions/3.0.2/lib/ruby/gems/3.0.0/gems/rack-test-1.1.0/lib/rack/test.rb:119:in `request'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/testing/integration.rb:279:in `process'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/testing/integration.rb:16:in `get'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/testing/integration.rb:370:in `get'
    /Users/yahonda/src/github.com/rails/rails/actionpack/test/dispatch/session/cookie_store_test.rb:395:in `get'
    /Users/yahonda/src/github.com/rails/rails/actionpack/test/dispatch/session/cookie_store_test.rb:170:in `block (2 levels) in test_deserializes_unloaded_classes_on_get_id'
    /Users/yahonda/src/github.com/rails/rails/actionpack/test/abstract_unit.rb:167:in `with_autoload_path'
    /Users/yahonda/src/github.com/rails/rails/actionpack/test/dispatch/session/cookie_store_test.rb:168:in `block in test_deserializes_unloaded_classes_on_get_id'
    /Users/yahonda/src/github.com/rails/rails/actionpack/test/dispatch/session/cookie_store_test.rb:413:in `block in with_test_route_set'
    /Users/yahonda/src/github.com/rails/rails/actionpack/test/abstract_unit.rb:153:in `with_routing'
    /Users/yahonda/src/github.com/rails/rails/actionpack/test/dispatch/session/cookie_store_test.rb:399:in `with_test_route_set'
    /Users/yahonda/src/github.com/rails/rails/actionpack/test/dispatch/session/cookie_store_test.rb:167:in `test_deserializes_unloaded_classes_on_get_id'


bin/test test/dispatch/session/cookie_store_test.rb:166



Finished in 0.212116s, 4.7144 runs/s, 0.0000 assertions/s.
1 runs, 0 assertions, 0 failures, 1 errors, 0 skips
%
```

```ruby
% bin/test test/dispatch/session/cookie_store_test.rb -n test_deserializes_unloaded_classes_on_get_value
Run options: -n test_deserializes_unloaded_classes_on_get_value --seed 1471

# Running:

E

Error:
CookieStoreTest#test_deserializes_unloaded_classes_on_get_value:
ActionDispatch::Session::SessionRestoreError: Session contains objects whose class definition isn't available.
Remember to require the classes for all objects kept in the session.
(Original exception: uninitialized constant SessionAutoloadTest [NameError])

    /Users/yahonda/src/github.com/rails/rails/activesupport/lib/active_support/inflector/methods.rb:280:in `constantize'
    /Users/yahonda/src/github.com/rails/rails/activesupport/lib/active_support/core_ext/string/inflections.rb:74:in `constantize'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/middleware/session/abstract_store.rb:58:in `rescue in stale_session_check!'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/middleware/session/abstract_store.rb:52:in `stale_session_check!'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/middleware/session/cookie_store.rb:88:in `block in unpacked_cookie_data'
    /Users/yahonda/.rbenv/versions/3.0.2/lib/ruby/gems/3.0.0/gems/rack-2.2.3/lib/rack/request.rb:69:in `fetch'
    /Users/yahonda/.rbenv/versions/3.0.2/lib/ruby/gems/3.0.0/gems/rack-2.2.3/lib/rack/request.rb:69:in `fetch_header'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/middleware/session/cookie_store.rb:87:in `unpacked_cookie_data'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/middleware/session/cookie_store.rb:81:in `block in extract_session_id'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/middleware/session/abstract_store.rb:53:in `stale_session_check!'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/middleware/session/cookie_store.rb:80:in `extract_session_id'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/request/session.rb:65:in `block in id'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/request/session.rb:64:in `fetch'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/request/session.rb:64:in `id'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/request/session.rb:84:in `id'
    /Users/yahonda/.rbenv/versions/3.0.2/lib/ruby/gems/3.0.0/gems/rack-2.2.3/lib/rack/session/abstract/id.rb:329:in `current_session_id'
    /Users/yahonda/.rbenv/versions/3.0.2/lib/ruby/gems/3.0.0/gems/rack-2.2.3/lib/rack/session/abstract/id.rb:335:in `session_exists?'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/request/session.rb:223:in `exists?'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/request/session.rb:246:in `load_for_read!'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/request/session.rb:111:in `[]'
    /Users/yahonda/src/github.com/rails/rails/actionpack/test/dispatch/session/cookie_store_test.rb:35:in `get_session_value'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_controller/metal/basic_implicit_render.rb:6:in `send_action'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/abstract_controller/base.rb:214:in `process_action'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_controller/metal/rendering.rb:53:in `process_action'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/abstract_controller/callbacks.rb:221:in `block in process_action'
    /Users/yahonda/src/github.com/rails/rails/activesupport/lib/active_support/callbacks.rb:99:in `run_callbacks'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/abstract_controller/callbacks.rb:220:in `process_action'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_controller/metal/rescue.rb:22:in `process_action'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_controller/metal/instrumentation.rb:65:in `block in process_action'
    /Users/yahonda/src/github.com/rails/rails/activesupport/lib/active_support/notifications.rb:207:in `block in instrument'
    /Users/yahonda/src/github.com/rails/rails/activesupport/lib/active_support/notifications/instrumenter.rb:24:in `instrument'
    /Users/yahonda/src/github.com/rails/rails/activesupport/lib/active_support/notifications.rb:207:in `instrument'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_controller/metal/instrumentation.rb:64:in `process_action'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_controller/metal/params_wrapper.rb:250:in `process_action'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/abstract_controller/base.rb:151:in `process'
    /Users/yahonda/src/github.com/rails/rails/actionview/lib/action_view/rendering.rb:39:in `process'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_controller/metal.rb:188:in `dispatch'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_controller/metal.rb:251:in `dispatch'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/routing/route_set.rb:49:in `dispatch'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/routing/route_set.rb:32:in `serve'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/journey/router.rb:50:in `block in serve'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/journey/router.rb:32:in `each'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/journey/router.rb:32:in `serve'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/routing/route_set.rb:845:in `call'
    /Users/yahonda/.rbenv/versions/3.0.2/lib/ruby/gems/3.0.0/gems/rack-2.2.3/lib/rack/session/abstract/id.rb:266:in `context'
    /Users/yahonda/.rbenv/versions/3.0.2/lib/ruby/gems/3.0.0/gems/rack-2.2.3/lib/rack/session/abstract/id.rb:260:in `call'
    /Users/yahonda/.rbenv/versions/3.0.2/lib/ruby/gems/3.0.0/gems/rack-2.2.3/lib/rack/head.rb:12:in `call'
    /Users/yahonda/.rbenv/versions/3.0.2/lib/ruby/gems/3.0.0/gems/rack-2.2.3/lib/rack/method_override.rb:24:in `call'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/middleware/cookies.rb:693:in `call'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/middleware/callbacks.rb:27:in `block in call'
    /Users/yahonda/src/github.com/rails/rails/activesupport/lib/active_support/callbacks.rb:99:in `run_callbacks'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/middleware/callbacks.rb:26:in `call'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/middleware/actionable_exceptions.rb:17:in `call'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/middleware/debug_exceptions.rb:28:in `call'
    /Users/yahonda/src/github.com/rails/rails/actionpack/test/abstract_unit.rb:96:in `call'
    /Users/yahonda/.rbenv/versions/3.0.2/lib/ruby/gems/3.0.0/gems/rack-test-1.1.0/lib/rack/mock_session.rb:29:in `request'
    /Users/yahonda/.rbenv/versions/3.0.2/lib/ruby/gems/3.0.0/gems/rack-test-1.1.0/lib/rack/test.rb:266:in `process_request'
    /Users/yahonda/.rbenv/versions/3.0.2/lib/ruby/gems/3.0.0/gems/rack-test-1.1.0/lib/rack/test.rb:119:in `request'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/testing/integration.rb:279:in `process'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/testing/integration.rb:16:in `get'
    /Users/yahonda/src/github.com/rails/rails/actionpack/lib/action_dispatch/testing/integration.rb:370:in `get'
    /Users/yahonda/src/github.com/rails/rails/actionpack/test/dispatch/session/cookie_store_test.rb:395:in `get'
    /Users/yahonda/src/github.com/rails/rails/actionpack/test/dispatch/session/cookie_store_test.rb:181:in `block (2 levels) in test_deserializes_unloaded_classes_on_get_value'
    /Users/yahonda/src/github.com/rails/rails/actionpack/test/abstract_unit.rb:167:in `with_autoload_path'
    /Users/yahonda/src/github.com/rails/rails/actionpack/test/dispatch/session/cookie_store_test.rb:179:in `block in test_deserializes_unloaded_classes_on_get_value'
    /Users/yahonda/src/github.com/rails/rails/actionpack/test/dispatch/session/cookie_store_test.rb:413:in `block in with_test_route_set'
    /Users/yahonda/src/github.com/rails/rails/actionpack/test/abstract_unit.rb:153:in `with_routing'
    /Users/yahonda/src/github.com/rails/rails/actionpack/test/dispatch/session/cookie_store_test.rb:399:in `with_test_route_set'
    /Users/yahonda/src/github.com/rails/rails/actionpack/test/dispatch/session/cookie_store_test.rb:178:in `test_deserializes_unloaded_classes_on_get_value'


bin/test test/dispatch/session/cookie_store_test.rb:177



Finished in 0.221102s, 4.5228 runs/s, 0.0000 assertions/s.
1 runs, 0 assertions, 0 failures, 1 errors, 0 skips
%
```

### System configuration
**Rails version**: main branch since https://github.com/rails/rails/pull/43028

**Ruby version**:ruby 3.0.2p107 (2021-07-07 revision 0db68f0233) [x86_64-darwin21]

**Comment by yahonda** [comment]
Opened #43035 to address these errors.

**Comment by kamipo** [comment]
Fixed by e933dc0b847ea67510c5330a22c2b8c4ab7bbc1a.
